Multiple usability improvements

adam-urbanczyk opened the pull request:
WIP. This PR will resolve #38 #33 #34 #23 #25 

- [x] Prompt on exit
- [x] Persist last file
- [x] Keyboard shortcuts
- [x] Handle color properly after unhiding
- [x] Fix save filter for STL
- [x] Do not scroll when saving with auto-reload
- [x] Object highlighting
- [x] Multi-selection
- [x] Better (multi-object) STEP and STL export

adam-urbanczyk commented:
@jmwright @rowanG077  I think I am getting there. Two-way graphical selection and multi object export (both STEP and STL) seem to be working. Still need to write tests, but it'd be great to hear some user feedback.

jmwright commented:
@adam-urbanczyk This is great, thanks for implementing this feature set.

Here's a little feedback (Ubuntu 18.04 64-bit).
* If I uncheck the model in the _Objects_  CQ models tree and then recheck it, the model in the 3D viewer comes back as a different color. Executing the script again reverts the model back to its original color.

![Screenshot_2019-04-11_07-46-57](https://user-images.githubusercontent.com/1015439/55954925-02fe9300-5c2e-11e9-94a1-20f3cc2721f8.png)

![Screenshot_2019-04-11_07-47-39](https://user-images.githubusercontent.com/1015439/55954946-19a4ea00-5c2e-11e9-95d0-a84ebcf09761.png)

* I can unselect the model by clicking in an open space in the 3D viewer, but I can't unselect the model by clicking in the open white space in the _Objects_ view. Not a big deal, but I instinctively tried to do it during testing.
* When exporting an STL file, stl is selected in the file type, but I still have to type the .stl extension onto the filename or the file saves without an extension. That prevents my slicer software from being able to see the file. Again, not a huge deal, but it improves the UX when the software helps the user along by figuring out whether or not it needs to tack on the file extension.
* When saving over an existing STL (from a previous export), CQ-editor asks whether you want to replace the file. However, even if you click _Yes_ it doesn't seem to overwrite the existing file with the new STL.
* STL export with multiple models selected seems to work fine.
* I like the visual differences in the 3D viewer between the three levels of selection: un-selected, mouse hover, selected. Not all software does that and I appreciate when it's present.
* I have not figured out how to do STEP export from the menu yet. I expected to see an _Export STEP_ option in the _Tools_ menu, but don't see it.

adam-urbanczyk commented:
Thanks for the feedback @jmwright ! It is very easy for me to miss those things.

* I fixed the color issue.
* Automatic adding of extensions seems to be working
* Added step export to the menu. NB: context menu on the object tree also works
* I understand your request regarding deselection in the object tree. It is however the default PyQT behavior. I won't touch it for now.

What are your thoughts in general about releasing CQ-editor? I'd still like to add keyboard shortcuts, but otherwise I think it is good enough for 0.1.  It would be also nice to sync it with CQ2 release

jmwright commented:
> What are your thoughts in general about releasing CQ-editor? I'd still like to add keyboard shortcuts, but otherwise I think it is good enough for 0.1. It would be also nice to sync it with CQ2 release

@adam-urbanczyk We're still blocked on Mac OS support, correct? It would be nice to get that figured out before release, if possible.

adam-urbanczyk commented:
Indeed would be nice but I don't think I will be able to solve it on my own (due to lack of  HW or VM).

jmwright commented:
Ok, I think releasing it is fine then. I keep wanting to get some concentrated time to sit down with CQ-editor on a Mac to troubleshoot, but haven't been able to make it happen.

One other thing that I've noticed is that the application won't prompt you to save your changes before exiting. With no File->Quit option I'm not sure how practical that would be to implement, but it's another thing I noticed.

adam-urbanczyk commented:
Good one @jmwright . That and persisting the last open file would be a nice addition!

rowanG077 commented:
This is great! I don;t really have any other comment besides that it work :)!

adam-urbanczyk commented:
Tests are fairly complete now - merging.

NB: large coverage increase is related to proper restoring of the tracing function in the debugger and trackback tests.
